pkg/services/sqlstore/migrations/accesscontrol/team_membership.go

```diff
@@ -0,0 +1,161 @@
+package accesscontrol
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"xorm.io/xorm"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+func AddTeamMembershipMigrations(mg *migrator.Migrator) {
+	mg.AddMigration("teams permissions migration", &teamPermissionMigrator{})
+}
+
+var _ migrator.CodeMigration = new(teamPermissionMigrator)
+
+type teamPermissionMigrator struct {
+	migrator.MigrationBase
+}
+
+func (p *teamPermissionMigrator) SQL(dialect migrator.Dialect) string {
+	return "code migration"
+}
+
+func (p *teamPermissionMigrator) Exec(sess *xorm.Session, migrator *migrator.Migrator) error {
+	return p.migrateMemberships(sess)
+}
+
+// generateNewRoleUID tries to generate a unique Short UID three times, in case of three conflicts returns an error.
+func generateNewRoleUID(sess *xorm.Session, orgID int64) (string, error) {
+	for i := 0; i < 3; i++ {
+		uid := util.GenerateShortUID()
+
+		exists, err := sess.Where("org_id=? AND uid=?", orgID, uid).Get(&accesscontrol.Role{})
+		if err != nil {
+			return "", err
+		}
+
+		if !exists {
+			return uid, nil
+		}
+	}
+
+	return "", fmt.Errorf("could not generate random uid")
+}
+
+// saveRole either creates a role or updates it with additional permissions
+func (p *teamPermissionMigrator) saveRole(sess *xorm.Session, orgID int64, name string, permissions []accesscontrol.Permission) (int64, bool, error) {
+	role := &accesscontrol.Role{OrgID: orgID, Name: name}
+	has, err := sess.Where("org_id = ? AND name = ?", orgID, name).Get(role)
+	if err != nil {
+		return 0, false, err
+	}
+
+	now := time.Now()
+	if !has {
+		uid, err := generateNewRoleUID(sess, orgID)
+		if err != nil {
+			return 0, false, err
+		}
+		role = &accesscontrol.Role{
+			UID:     uid,
+			Name:    name,
+			OrgID:   orgID,
+			Updated: now,
+			Created: now,
+		}
+		if _, err := sess.Insert(role); err != nil {
+			return 0, false, err
+		}
+	}
+
+	var newPermissions []accesscontrol.Permission
+	for _, permission := range permissions {
+		has, err := sess.Where("role_id = ? AND action = ? AND scope = ?", role.ID, permission.Action, permission.Scope).Get(&accesscontrol.Permission{})
+		if err != nil {
+			return 0, false, err
+		}
+		if !has {
+			permission.RoleID = role.ID
+			permission.Created = now
+			permission.Updated = now
+			newPermissions = append(newPermissions, permission)
+		}
+	}
+
+	_, err = sess.InsertMulti(&newPermissions)
+	if err != nil {
+		return 0, false, err
+	}
+
+	return role.ID, has, nil
+}
+
+// mapPermissionToFGAC translates the legacy membership (Member or Admin) into FGAC permissions
+func (p *teamPermissionMigrator) mapPermissionToFGAC(permission models.PermissionType, teamID int64) []accesscontrol.Permission {
+	teamIDScope := accesscontrol.Scope("teams", "id", strconv.FormatInt(teamID, 10))
+	switch permission {
+	case 0:
```

Comment:
nit: is there a constant for this? `PERMISSION_VIEWER`?

Reply:
Unfortunately not 😢  For some reason `member` permission is not using the enums as instead just stored as 0. This has caused some issues elsewhere as well (for instance, [here](https://github.com/grafana/grafana/pull/43951/files#diff-3d63f0ed19b7a13f6d51ed456264cf47561a2dcfb1486911c114df6ce3c5a133R125-R133)), but I didn't want to change the legacy behaviour, so...